pepdbagent/search.py

```diff
@@ -0,0 +1,257 @@
+                        submission_date=str(res[7]),
+                    )
+                )
+        except KeyError:
+            results_list = []
+
+        return results_list
+
+    def _get_count_of_found_namespaces(
+        self, search_str: str = "", admin_nsp: tuple = None
+    ) -> int:
+        """
+        Get number of found namespace. [This function is related to _find_namespaces]
+        :param search_str: string of symbols, words, keywords to search in the
+            namespace name.
+        :param admin_nsp: tuple of namespaces where project can be retrieved if they are privet
+        :return: number of found namespaces
+        """
+        count_sql = f"""
+        select COUNT(DISTINCT ({NAMESPACE_COL}))
+            from {DB_TABLE_NAME} where (({NAMESPACE_COL} ILIKE '%%{search_str}%%' and {PRIVATE_COL} = false)
+                or ({NAMESPACE_COL} ILIKE '%%{search_str}%%' and {NAMESPACE_COL} in %s )) 
+        """
+        result = self.__run_sql_fetchall(count_sql, admin_nsp)
+        try:
+            number_of_prj = result[0][0]
+        except KeyError:
+            number_of_prj = 0
+        return number_of_prj
+
+    def _find_namespaces(
+        self,
+        search_str: str,
+        admin_nsp: tuple = None,
+        limit: int = DEFAULT_LIMIT,
+        offset: int = DEFAULT_OFFSET,
+    ) -> List[NamespaceSearchResultModel]:
+        """
+        Search for namespace by providing search string.
+        :param search_str: string of symbols, words, keywords to search in the
+            namespace name.
+        :param admin_nsp: tuple of namespaces where project can be retrieved if they are privet
+        :param limit: limit of return results
+        :param offset: number of results off set (that were already showed)
+        :return: list of dict with strucutre {
+                namespace,
+                number_of_projects,
+                number_of_samples,
+            }
+        """
+        count_sql = f"""
+        select {NAMESPACE_COL}, COUNT({NAME_COL}), SUM({N_SAMPLES_COL})
+            from {DB_TABLE_NAME} where (({NAMESPACE_COL} ILIKE '%%{search_str}%%' and {PRIVATE_COL} is false)
+                or ({NAMESPACE_COL} ILIKE '%%{search_str}%%' and {NAMESPACE_COL} in %s )) 
+                    GROUP BY {NAMESPACE_COL}
+                        LIMIT {limit} OFFSET {offset};
+        """
+        results = self.__run_sql_fetchall(count_sql, admin_nsp)
+        results_list = []
+        try:
+            for res in results:
+                results_list.append(
+                    NamespaceSearchResultModel(
+                        namespace=res[0],
+                        number_of_projects=res[1],
+                        number_of_samples=res[2],
+                    )
+                )
+        except KeyError:
+            results_list = []
+
+        return results_list
+
+    def __run_sql_fetchall(self, sql_query: str, *argv) -> list:
```

Comment:
Might be over-engineering, but I notice that you use this same method in the `Connection` class. Might be worth making a `BaseConnection` class that has these sort of methods, and let `Connection` and `Search` inherit from it

Reply:
I didn't know how to solve this problem, but now you gave me an idea, thanks!!
I try to write something like that